torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -133,36 +184,40 @@ TensorPipeAgent::~TensorPipeAgent() {
 
 void TensorPipeAgent::startImpl() {
   VLOG(1) << "RPC agent for " << workerInfo_.name_ << " is starting";
-  auto uvContext = std::make_shared<tensorpipe::transport::uv::Context>();
 
-  std::string uvAddress = guessAddress(*uvContext);
-  VLOG(1) << "RPC agent for " << workerInfo_.name_ << " is using address "
-          << uvAddress;
+  std::vector<std::string> addresses;
+  int lowestPriority = -1;
+  std::string lowestPriorityTransport;
 
-  context_->registerTransport(1, "tcp", std::move(uvContext));
-#ifdef TP_ENABLE_SHM
-  context_->registerTransport(
-      0, "shm", std::make_shared<tensorpipe::transport::shm::Context>());
-#endif
-  context_->registerChannel(
-      1, "basic", std::make_shared<tensorpipe::channel::basic::Context>());
-#ifdef TP_ENABLE_CMA
-  context_->registerChannel(
-      0, "cma", std::make_shared<tensorpipe::channel::cma::Context>());
-#endif
+  for (auto& key : TensorPipeTransportRegistry()->Keys()) {
```

Comment:
I assume the keys here will be tcp/shm/cma? What if I want to add two tcp channels for two different NICs?

Reply:
Indeed, the keys at the moment are ones of the registry, which means that for now only one instance of each transport/channel is allowed, and thus we cannot have two TCP channels with different configurations. This is one of the things we need to figure out, probably by v1.7, related to the agent options. For example, we could make it so that the transports/channels in registry are "templates" that the user can instantiate multiple times if desired, like:
```
options=rpc.TensorPipeAgentOptions(transports={
    "tcp_nic0": Transport("uv", device="eth0"),
    "tcp_nic1": Transport("uv", device="eth1"),
})
```